client/cli/src/config.rs

```diff
@@ -0,0 +1,308 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Configuration trait for a CLI based on substrate
+
+use crate::error::Result;
+use crate::SubstrateCLI;
+use app_dirs::{AppDataType, AppInfo};
+use names::{Generator, Name};
+use sc_service::config::{
+	Configuration, DatabaseConfig, ExecutionStrategies, ExtTransport, KeystoreConfig,
+	NetworkConfiguration, NodeKeyConfig, PrometheusConfig, PruningMode, Roles, TelemetryEndpoints,
+	TransactionPoolOptions, WasmExecutionMethod,
+};
+use sc_service::ChainSpec;
+use std::future::Future;
+use std::net::SocketAddr;
+use std::path::PathBuf;
+use std::pin::Pin;
+use std::sync::Arc;
+
+/// The maximum number of characters for a node name.
+pub(crate) const NODE_NAME_MAX_LENGTH: usize = 32;
+
+/// default sub directory to store network config
+pub(crate) const DEFAULT_NETWORK_CONFIG_PATH: &'static str = "network";
+
+/// A trait that allows converting an object to a Configuration
+pub trait CliConfiguration: Sized {
+	/// Get the base path of the configuration (if any)
+	fn get_base_path(&self) -> Result<Option<&PathBuf>>;
+
+	/// Returns `true` if the node is for development or not
+	fn get_is_dev(&self) -> Result<bool> {
+		Ok(false)
+	}
+
+	/// Get the roles
+	fn get_roles(&self, _is_dev: bool) -> Result<Roles> {
+		Ok(Roles::FULL)
+	}
+
+	/// Get the transaction pool options
+	fn get_transaction_pool(&self) -> Result<TransactionPoolOptions> {
```

Comment:
Hmm, this method name doesn't match return type.

You could either
1. rename this to `get_transaction_pool_options`
2. drop the `get_` prefix so you don't have long-ass method names.

Reply:
I will drop the `get_` everywhere as suggested here: https://github.com/paritytech/substrate/pull/5271#issuecomment-602550473